airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/common.py

```diff
@@ -0,0 +1,155 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+import json
+from abc import ABC, abstractmethod
+from http import HTTPStatus
+from typing import Any, Iterable, Mapping, MutableMapping, Optional
+
+import requests
+from airbyte_cdk.sources.streams.core import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from pydantic import BaseModel, ValidationError
+from source_amazon_ads.common import PageToken, SourceContext
+from source_amazon_ads.schemas import JSModel
+
+URL_BASE = "https://advertising-api.amazon.com/"
+
+
+class ErrorResponse(BaseModel):
+    code: str
+    details: str
+    requestId: str
+
+
+class BasicAmazonAdsStream(Stream, ABC):
+    def __init__(self, config, context: SourceContext = None):
+        self._ctx = context or SourceContext()
+        self._config = config
+        self._url = self._config.host or URL_BASE
+
+    @property
+    @abstractmethod
+    def model(self) -> JSModel:
+        """
+        Pydantic model to represent json schema
+        """
+
+    def get_json_schema(self):
+        return self.model.schema()
+
+
+# Basic full refresh stream
+class AmazonAdsStream(HttpStream, BasicAmazonAdsStream):
+    def __init__(self, config, *args, context: SourceContext = None, **kwargs):
+        BasicAmazonAdsStream.__init__(self, config, context=context or SourceContext())
+        HttpStream.__init__(self, *args, **kwargs)
+
+    @property
+    def url_base(self):
+        return self._url
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_headers(self, *args, **kvargs) -> MutableMapping[str, Any]:
+        return {"Amazon-Advertising-API-ClientId": self._config.client_id}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an object representing single record in the response
+        """
+        yield from json.loads(response.text) if response else []
+
+    def _send_request(self, request: requests.PreparedRequest, request_kwargs: Mapping[str, Any]) -> requests.Response:
+        try:
+            return super()._send_request(request, request_kwargs)
+        except requests.exceptions.HTTPError as http_exception:
+            response = http_exception.response
+            if response.status_code in [HTTPStatus.FORBIDDEN]:
+                try:
+                    resp = ErrorResponse.parse_raw(response.text)
+                    self.logger.warn(
+                        f"Unexpected error {resp.code} when processing request {request.url} for {request.headers['Amazon-Advertising-API-Scope']} profile: {resp.details}"
+                    )
+                except ValidationError:
+                    raise http_exception
```

Comment:
If we receive a 403 error, then you do not want to interrupt the reading, but simply display the log, but if the error does not fit into your ErrorModel, then you call the exception.
I think the logic is a little flawed.

Reply:
Yeah,  thats how Amazon Ads API server works. Ive added this logic after faced with problem getting display campaigns for one of the seller profile it was returning 403 with this json:
 ```
{
    "code": "403",
    "details": "Advertiser is not authorized for Sponsored Display.",
    "requestId": "Y1MSNQKCHZ8AQ9QXKEVV"
}
```
It was only for one seller account and only for campaign stream. Despite it has 403 response code we don't want this error to stop processing another streams (cause every  sync  run would result in error) unlike regular wrong auth token error that looks like this:
```
{
    "message": "Missing Authentication Token"
}
``` 
So  only way to make a difference between critical errors (wrong/expired auth token) and non-critical error like permission issues is to check if json fits into format described by the API reference.

Reply:
not sure why we raising exceptions (i.e. stoping our read) in case we are unable to parse the error.